aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -504,30 +504,48 @@ Tensor trace_cpu(const Tensor& self) {
   Tensor result;
   ScalarType dtype = get_dtype(result, self, c10::nullopt, true);
   result = at::empty({}, self.options().dtype(dtype));
-  AT_DISPATCH_ALL_TYPES(self.scalar_type(), "trace", [&] {
-    using accscalar_t = at::acc_type<scalar_t, false>;
-    accscalar_t sum = 0;
-    const auto* t_data = self.data_ptr<scalar_t>();
+  if (result.scalar_type() == at::kLong) {
+    AT_DISPATCH_INTEGRAL_TYPES(self.scalar_type(), "trace", [&] {
+      using accscalar_t = at::acc_type<scalar_t, false>;
+      accscalar_t sum = 0;
+      const auto* t_data = self.data_ptr<scalar_t>();
 
-    int64_t t_stride_0, t_stride_1, t_diag_size;
+      int64_t t_stride_0, t_stride_1, t_diag_size;
 
-    TORCH_CHECK(self.dim() == 2, "trace: expected a matrix, but got tensor with dim ", self.dim());
+      TORCH_CHECK(self.dim() == 2, "trace: expected a matrix, but got tensor with dim ", self.dim());
 
-    t_stride_0 = self.stride(0);
-    t_stride_1 = self.stride(1);
+      t_stride_0 = self.stride(0);
+      t_stride_1 = self.stride(1);
 
-    t_diag_size = std::min(self.size(0), self.size(1));
-    for (int64_t i = 0; i < t_diag_size; i++) {
-      sum += t_data[i * (t_stride_0 + t_stride_1)];
-    }
+      t_diag_size = std::min(self.size(0), self.size(1));
+      for (int64_t i = 0; i < t_diag_size; i++) {
+        sum += t_data[i * (t_stride_0 + t_stride_1)];
+      }
 
-    // all integer types get promoted to kLong
-    if (result.scalar_type() == at::kLong) {
+      // all integer types get promoted to kLong
       *result.data_ptr<int64_t>() = sum;
```

Comment:
separated dispatch because this cast causes a compile error for complex otherwise

Reply:
This seems like a good opportunity to use a `if_constexpr` to fix the problem. Test for `std::is_integral<scalar_t>` . Check `if_constexpr` docs for how to use it.

Reply:
still gives a compile error